examples/naive/aarch64/intt_dilithium_1234_5678_manual_ld4.s

```diff
@@ -0,0 +1,512 @@
+        ldr_vo root3, \r_ptr, (-8*16 + 3*16)
+        ldr_vo root4, \r_ptr, (-8*16 + 4*16)
+        ldr_vo root5, \r_ptr, (-8*16 + 5*16)
+        ldr_vo root6, \r_ptr, (-8*16 + 6*16)
+        ldr_vo root7, \r_ptr, (-8*16 + 7*16)
+.endm
+
+.macro load_next_roots_56 root0, r_ptr0
+        ldr_vi \root0, \r_ptr0, 16
+.endm
+
+.macro load_next_roots_6 root0, r_ptr0
+        ldr_vi \root0, \r_ptr0, 8
+.endm
+
+.macro load_next_roots_78 root0, root0_tw, root1, root1_tw, root2, root2_tw, r_ptr1
+        ldr_vi \root0,    \r_ptr1, (6*16)
+        ldr_vo \root0_tw, \r_ptr1, (-6*16 + 1*16)
+        ldr_vo \root1,    \r_ptr1, (-6*16 + 2*16)
+        ldr_vo \root1_tw, \r_ptr1, (-6*16 + 3*16)
+        ldr_vo \root2,    \r_ptr1, (-6*16 + 4*16)
+        ldr_vo \root2_tw, \r_ptr1, (-6*16 + 5*16)
+.endm
+
+.macro transpose4 data
+        trn1_s t0, \data\()0, \data\()1
+        trn2_s t1, \data\()0, \data\()1
+        trn1_s t2, \data\()2, \data\()3
+        trn2_s t3, \data\()2, \data\()3
+
+        trn2_d \data\()2, t0, t2
+        trn2_d \data\()3, t1, t3
+        trn1_d \data\()0, t0, t2
+        trn1_d \data\()1, t1, t3
+.endm
+
+.macro save_gprs // @slothy:no-unfold
+        sub sp, sp, #(16*6)
+        stp x19, x20, [sp, #16*0]
+        stp x19, x20, [sp, #16*0]
+        stp x21, x22, [sp, #16*1]
+        stp x23, x24, [sp, #16*2]
+        stp x25, x26, [sp, #16*3]
+        stp x27, x28, [sp, #16*4]
+        str x29, [sp, #16*5]
+.endm
+
+.macro restore_gprs // @slothy:no-unfold
+        ldp x19, x20, [sp, #16*0]
+        ldp x21, x22, [sp, #16*1]
+        ldp x23, x24, [sp, #16*2]
+        ldp x25, x26, [sp, #16*3]
+        ldp x27, x28, [sp, #16*4]
+        ldr x29, [sp, #16*5]
+        add sp, sp, #(16*6)
+.endm
+
+.macro save_vregs // @slothy:no-unfold
+        sub sp, sp, #(16*4)
+        stp  d8,  d9, [sp, #16*0]
+        stp d10, d11, [sp, #16*1]
+        stp d12, d13, [sp, #16*2]
+        stp d14, d15, [sp, #16*3]
+.endm
+
+.macro restore_vregs // @slothy:no-unfold
+        ldp  d8,  d9, [sp, #16*0]
+        ldp d10, d11, [sp, #16*1]
+        ldp d12, d13, [sp, #16*2]
+        ldp d14, d15, [sp, #16*3]
+        add sp, sp, #(16*4)
+.endm
+
+#define STACK_SIZE 16
+#define STACK0 0
+
+.macro restore a, loc     // @slothy:no-unfold
+        ldr \a, [sp, #\loc\()]
+.endm
+.macro save loc, a        // @slothy:no-unfold
+        str \a, [sp, #\loc\()]
+.endm
+.macro push_stack // @slothy:no-unfold
+        save_gprs
+        save_vregs
+        sub sp, sp, #STACK_SIZE
+.endm
+
+.macro pop_stack // @slothy:no-unfold
+        add sp, sp, #STACK_SIZE
+        restore_vregs
+        restore_gprs
+.endm
+
+.data
+.p2align 4
+roots:
+#include "intt_dilithium_1234_5678_twiddles.s"
+.text
+
+        .global intt_dilithium_1234_5678_manual_ld4
+        .global _intt_dilithium_1234_5678_manual_ld4
+
+.p2align 4
+modulus_addr:   .quad 8380417
+ninv_addr:      .quad 16382
+ninv_tw_addr:   .quad 4197891
+intt_dilithium_1234_5678_manual_ld4:
```

Comment:
Again an optional improvement, but better to do while we're at it:

Could you go through the clean [inv]NTTs and add a comment at their entry symbol describing their arguments and in what way (reduction / ordering) they deviate from a standard, fully-reduced NTT?

Reply:
That's a good point in general. I followed the way the current PQClean implementations handle the reductions sucht that we can have a fair comparison. 

Regarding the ordering: All of our NTTs on aarch64 (should) output in the canonical ordering.